File: subprojects/ide-plugin/src/main/java/org/gradle/profiler/studio/plugin/system/IdeSystemHelper.java
package org.gradle.profiler.studio.plugin.system;

import com.google.common.base.Strings;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.externalSystem.importing.ImportSpecBuilder;
import com.intellij.openapi.externalSystem.util.ExternalSystemUtil;
import com.intellij.openapi.progress.ProgressModel;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupManager;
import com.intellij.openapi.wm.IdeFrame;
import com.intellij.openapi.wm.ex.StatusBarEx;
import com.intellij.openapi.wm.ex.WindowManagerEx;
import org.gradle.profiler.client.protocol.messages.IdeSyncRequestCompleted.IdeSyncRequestResult;
import org.jetbrains.plugins.gradle.util.GradleConstants;

import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;

public class IdeSystemHelper {

    private static final long WAIT_ON_PROCESS_SLEEP_TIME = 10;
    private static final long WAIT_ON_STARTUP_SLEEP_TIME = 100;

    /**
     * Gets the result of startup sync by checking if there was already any sync done before. Otherwise, it starts a new sync.
     */
    public static GradleSyncResult getStartupSyncResult(Project project, GradleSystemListener gradleSystemListener) {
        if (!gradleSystemListener.hasAnySyncCompleted()) {
            // Sync was not run before, we need to run it manually
            GradleSyncResult result = startManualSync(project, gradleSystemListener);
            if (result.getResult() == IdeSyncRequestResult.FAILED) {
                // If it fails, it might fail because another sync just started a millisecond before we could start it
                waitOnPreviousGradleSyncFinish(gradleSystemListener);
                waitOnBackgroundProcessesFinish(project);
            }
        }
        return buildSyncResult(gradleSystemListener);
    }

    /**
     * Starts a manual sync and returns a result.
     */
    public static GradleSyncResult startManualSync(Project project, GradleSystemListener gradleSystemListener) {
        CompletableFuture<Void> nextSyncDone = gradleSystemListener.awaitNextSyncCompletion();
        ExternalSystemUtil.refreshProjects(
            new ImportSpecBuilder(project, GradleConstants.SYSTEM_ID)
        );
        nextSyncDone.join();
        waitOnBackgroundProcessesFinish(project);
        return buildSyncResult(gradleSystemListener);
    }

    private static GradleSyncResult buildSyncResult(GradleSystemListener gradleSystemListener) {
        Exception exception = gradleSystemListener.getLastException();
        if (exception != null) {
            return new GradleSyncResult(IdeSyncRequestResult.FAILED, Strings.nullToEmpty(exception.getMessage()));
        }
        return new GradleSyncResult(IdeSyncRequestResult.SUCCEEDED, "");
    }

    /**
     * Waits for any in-progress Gradle sync to finish.
     */
    public static void waitOnPreviousGradleSyncFinish(GradleSystemListener gradleSystemListener) {
        gradleSystemListener.waitForCurrentSyncToFinish();
    }

    /**
     * Wait on IDE indexing and similar background tasks to finish.
     * <p>
     * It seems there is no better way to do it atm.
     */
    public static void waitOnBackgroundProcessesFinish(Project project) {
        DumbService.getInstance(project).waitForSmartMode();
        IdeFrame frame = WindowManagerEx.getInstanceEx().findFrameFor(project);
        StatusBarEx statusBar = frame == null ? null : (StatusBarEx) frame.getStatusBar();
        if (statusBar != null) {
            statusBar.getBackgroundProcessModels().forEach(it -> waitOnProgressModel(it.getSecond()));
        }
    }

    private static void waitOnProgressModel(ProgressModel progressModel) {
        wait(WAIT_ON_PROCESS_SLEEP_TIME, progressModel::isRunning);
    }

    public static void waitOnPostStartupActivities(Project project) {
        wait(WAIT_ON_STARTUP_SLEEP_TIME, () -> !StartupManager.getInstance(project).postStartupActivityPassed());
    }

    @SuppressWarnings("BusyWait")
    private static void wait(long sleepMillis, Supplier<Boolean> whileCondition) {
        while (whileCondition.get()) {
            try {
                Thread.sleep(sleepMillis);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }

    /**
     * Exit the application.
     */
    public static void exit(Project project) {
        waitOnBackgroundProcessesFinish(project);
        ApplicationManager.getApplication().exit(true, true, false);
    }
}

File: subprojects/intellij-plugin/src/main/kotlin/org/gradle/profiler/intellij/plugin/system/IdeSystemHelper.kt
package org.gradle.profiler.intellij.plugin.system

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.StartupManager
import com.intellij.openapi.wm.ex.StatusBarEx
import com.intellij.openapi.wm.ex.WindowManagerEx

/**
 * Common IDE helper utilities that work across both Android Studio and IntelliJ IDEA.
 */
object IdeSystemHelper {

    private const val WAIT_ON_PROCESS_SLEEP_TIME = 10L
    private const val WAIT_ON_STARTUP_SLEEP_TIME = 100L

    /** Waits for IDE indexing and any visible background tasks to finish. */
    @JvmStatic
    fun waitOnBackgroundProcessesFinish(project: Project) {
        // A dummy read action in smart mode blocks until all indexing is done.
        DumbService.getInstance(project).runReadActionInSmartMode {}
        val frame = WindowManagerEx.getInstanceEx().findFrameFor(project)
        val statusBar = frame?.statusBar as? StatusBarEx
        statusBar?.backgroundProcesses?.forEach { waitOnProgressIndicator(it.second) }
    }

    private fun waitOnProgressIndicator(progressIndicator: ProgressIndicator) {
        wait(WAIT_ON_PROCESS_SLEEP_TIME) { progressIndicator.isRunning }
    }

    @JvmStatic
    fun waitOnPostStartupActivities(project: Project) {
        wait(WAIT_ON_STARTUP_SLEEP_TIME) { !StartupManager.getInstance(project).postStartupActivityPassed() }
    }

    /** Waits for background processes to finish then exits the application. */
    @JvmStatic
    fun exit(project: Project) {
        waitOnBackgroundProcessesFinish(project)
        ApplicationManager.getApplication().exit(true, true, false)
    }

    private fun wait(sleepMillis: Long, whileCondition: () -> Boolean) {
        while (whileCondition()) {
            Thread.sleep(sleepMillis)
        }
    }
}
